app/models/manageiq/providers/autosde/storage_manager/ems_refresh_workflow.rb

```diff
@@ -22,6 +28,26 @@ def poll_native_task
     signal(:abort, err.message, "error")
   end
 
+  def poll_refresh_task
+    opts = {
+      :query_params => {
+        :object_id => options[:native_object_id]
+      }
+    }
+    tasks = autosde_client.JobApi.jobs_get(opts)
+    case tasks.first.status
```

Comment:
I think you need to handle if tasks is `[]` also otherwise this will raise an exception.

Reply:
Is there a job-type or naming convention that we can use to identify the "refresh task" versus some other task?  What happens if another task for that object_id kicks off that finishes quickly before the self-refresh task is done?

Reply:
Good question. There is, and I actually wrote it at the beginning, but then I decided to remove it since it makes a tide coupling between autosde and miq. It's actually a name that could change very easily and break everything, so I didn't like the idea.

And then comes your question, what happens if another task finishes before that refresh is done. The answer is that it shouldn't happen, because we're speaking here about the creation of the storage, and the user can't really see anything until that first refresh task is complete. And even in an extreme scenario where that does happen for some reason, what will happen is that the refresh will wait for that new task, and not only I don't mind that but actually it's preferable, so that the refresh also watches after that change as well.

Reply:
> I think you need to handle if tasks is [] also otherwise this will raise an exception.

I don't understand, the whole function is handled, it will get to the abort... or do you mean to handle it individually and give it a more friendly error message?

Reply:
> > I don't understand, the whole function is handled, it will get to the abort... or do you mean to handle it individually and give it a more friendly error message?
> 
> I'm thinking if somehow we check before the task is created, this will blow up and abort as opposed to retry.
> 
> Plus if you want to require there to be exactly 1 result in an array it is better to check that and raise a failure with a proper error message rather than a confusing `NoMethodError first for NilClass`

Now I got it. But to be honest, I actually prefer it to abort rather than to retry. Because synchronously speaking, the autosde's refresh task is being fired before the create task is finished (and then it behaves asynchronously, but it the refresh task isn't saved, the autosde's refresh won't success).
So technically, this can't happen. And if for some strange reason it does happen, if the task wasn't created till then, it won't be created at all, so there is no reason to retry. I do agree with the proper error message. I don't think we need it because I don't see a scenario of that happening, but if you want I'll add it.

Reply:
Done. But I wrote < 1, because it's supposed to be 2 (it will also find the task of the creation of the storage system, which has the same object_id)

By the way, I changed the time of the delete of storage, it takes way less time that one minute, if it's Ok :)

Reply:
Okay highly discourage using `.first` on an array with 2 items without first sorting or selecting on some unique value.  Is there some attribute about the storage system creation task that we can use to exclude it or something about the refresh task we can select on?

`tasks.reject { |t| t.name.match?("Create Storage System") }.first`
or
`tasks.detect { |t| t.name.match("Refresh task") }`

Reply:
That's what I wanted to avoid in the first place, but now that you say, If we bring everything and then filter it here, it makes more sense. So we're looking for "update_system", that's not supposed to change.
So cool.
Do you want me to add:
`raise _('No refresh task found') if task == nil`
?

Reply:
Just checking, could you use `"update_system"` in your query_params?

Reply:
I can use "autosde_site_manager.engine.tasks.update_system" but that's what I told you that I wanted to avoid because of the coupling. And I definitely don't want to allow the REST to find something by part of its name by standard...